pkg/requestutil/request_info.go

```diff
@@ -0,0 +1,128 @@
+// Copyright 2021 TiKV Project Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package requestutil
+
+import (
+	"bytes"
+	"encoding/json"
+	"fmt"
+	"io"
+	"net/http"
+	"strings"
+	"time"
+
+	"github.com/tikv/pd/pkg/apiutil"
+)
+
+// RequestInfo holds source information from http.Request
+type RequestInfo struct {
+	ServiceLabel string
+	Method       string
+	Component    string
+	IP           string
+	TimeStamp    string
+	URLParam     string
+	BodyParm     string
+}
+
+// RequestSchemaList is used to store requestSchemas which have been registered
+type RequestSchemaList struct {
+	requestSchemas []*RequestSchema
+}
+
+// NewRequestSchemaList returns a new RequestSchemaList with given length
+func NewRequestSchemaList(len int) *RequestSchemaList {
+	return &RequestSchemaList{requestSchemas: make([]*RequestSchema, 0, len)}
+}
+
+func (l *RequestSchemaList) match(path string, method string) string {
```

Comment:
I wonder this may not be as efficient as hash method. especially if, APIs that are accessed frequently and register several services. may have performance issues. Can we add a  mirco benchmark?

Reply:
get it